Create Melt Reaction Modules

Grant-Block opened the pull request:
This is a work in progress PR! So far, I've made a new module called MeltModel with a new class that currently contains a function to calculate melt fraction, taken from the MeltSimple model. This was plugged back into MeltSimple to make sure everything works, and can hopefully soon be extended to the VP model

Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*

### Before your first pull request:

* [X ] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [ X] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ X] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

naliboff commented on include/aspect/material_model/melt_model/melt_simple_fraction.h:
```diff
@@ -0,0 +1,102 @@
+    namespace MeltModel
+    {
+      template <int dim>
+      class MeltSimpleFraction : public ::aspect::SimulatorAccess<dim>
```
My initial vote is to make this more descriptive, and have it names something like `Katz03MantleMelting` or something equally descriptive such that in can be distinguished from future distinct melting models (and other types of reactions) that are added.

naliboff commented on include/aspect/material_model/melt_model/melt_simple_fraction.h:
```diff
@@ -0,0 +1,102 @@
+           */
+          virtual
+          double
+          melt_fraction (const double temperature,
```
Note - for historical reasons, I think after discussion with @jdannberg we decided the name of this function should stay as `melt_fraction` even if dealing with non-melt fluids (water, etc).

naliboff commented on include/aspect/material_model/melt_model/melt_simple_fraction.h:
```diff
@@ -0,0 +1,102 @@
+          // melt fraction exponent
+          double beta;
+
+
```
Remove this line

naliboff commented on include/aspect/material_model/melt_simple.h:
```diff
@@ -171,15 +173,17 @@ namespace aspect
+        // virtual
+        // double
+        // melt_fraction (const double temperature,
+        //                const double pressure) const;
```
Is the original code?

Grant-Block replied:
This is the original code I left in there just to make sure I wasn't messing anything up during this change. I can get rid of it now!

naliboff reviewed (commented):
@Grant-Block - Thanks for taking a first stab at this! I think the first-order structure makes sense, but I have some suggestions for how to name the high-level folder, files, etc. Let's discuss with @jdannberg and then we can dive into the details.

gassmoeller commented on include/aspect/material_model/melt_model/melt_simple_fraction.h:
```diff
@@ -0,0 +1,102 @@
+           * @p pressure (assuming equilibrium conditions). Melting model after Katz,
+           * 2003, for dry peridotite.
+           */
+          virtual
```
Unless you expect this class to act as a base class there is no need to make this function `virtual`.

gassmoeller commented on include/aspect/material_model/melt_simple.h:
```diff
@@ -49,8 +50,9 @@ namespace aspect
      * and contains compressibility for both solid and melt.
      *
-     * @ingroup MaterialModels
+     * @ingroup MeltModel
      */
+    namespace MeltModel {}
```
I dont understand this change. The class `MeltSimple` is still a material model, not a melt model, so its documentation should stay in the `MaterialModels` group.

And opening and closing the namespace `MeltModel` will do nothing here. Please remove.

gassmoeller reviewed (commented):
I will leave the rest of the review for @jdannberg. Just a small comment from me.

jdannberg commented on source/material_model/melt_model/melt_simple_fraction.cc:
```diff
@@ -0,0 +1,205 @@
+
+
+#include <aspect/material_model/melt_model/melt_simple_fraction.h>
+#include <aspect/adiabatic_conditions/interface.h>
+#include <aspect/gravity_model/interface.h>
+#include <aspect/utilities.h>
+
+#include <deal.II/base/parameter_handler.h>
+#include <deal.II/numerics/fe_field_function.h>
```
```suggestion
#include <aspect/utilities.h>

#include <deal.II/base/parameter_handler.h>
```
I think you don't need these, right?

jdannberg commented on source/material_model/melt_simple.cc:
```diff
@@ -49,43 +50,43 @@ namespace aspect
-        }
-      return peridotite_melt_fraction;
-    }
+    // template <int dim>
+    // double
+    // MeltSimple<dim>::
+    // melt_fraction (const double temperature,
+    //                const double pressure) const
+    // {
+    //   // anhydrous melting of peridotite after Katz, 2003
+    //   const double T_solidus  = A1 + 273.15
+    //                             + A2 * pressure
+    //                             + A3 * pressure * pressure;
+    //   const double T_lherz_liquidus = B1 + 273.15
+    //                                   + B2 * pressure
+    //                                   + B3 * pressure * pressure;
+    //   const double T_liquidus = C1 + 273.15
+    //                             + C2 * pressure
+    //                             + C3 * pressure * pressure;
+
+    //   // melt fraction for peridotite with clinopyroxene
+    //   double peridotite_melt_fraction;
+    //   if (temperature < T_solidus || pressure > 1.3e10)
+    //     peridotite_melt_fraction = 0.0;
+    //   else if (temperature > T_lherz_liquidus)
+    //     peridotite_melt_fraction = 1.0;
+    //   else
+    //     peridotite_melt_fraction = std::pow((temperature - T_solidus) / (T_lherz_liquidus - T_solidus),beta);
+
+    //   // melt fraction after melting of all clinopyroxene
+    //   const double R_cpx = r1 + r2 * std::max(0.0, pressure);
+    //   const double F_max = M_cpx / R_cpx;
+
+    //   if (peridotite_melt_fraction > F_max && temperature < T_liquidus)
+    //     {
+    //       const double T_max = std::pow(F_max,1/beta) * (T_lherz_liquidus - T_solidus) + T_solidus;
+    //       peridotite_melt_fraction = F_max + (1 - F_max) * pow((temperature - T_max) / (T_liquidus - T_max),beta);
+    //     }
+    //   return peridotite_melt_fraction;
+    // }
```
Same here, you can now remove all the commented out code.

jdannberg commented on source/material_model/melt_simple.cc:
```diff
@@ -96,6 +97,7 @@ namespace aspect
                     const double maximum_melt_fraction,
                     const NonlinearDependence::Dependence dependence) const
     {
+
```
Can you remove this empty line again?

jdannberg commented on source/material_model/melt_simple.cc:
```diff
@@ -96,6 +97,7 @@ namespace aspect
+
       double entropy_gradient = 0.0;
 
       // calculate latent heat of melting
```
I do not understand how this still works, currently, since you're not reading in these parameters anymore. 
I think you need to move all of this into your melting model as well. 

This also means that apparently, we do not have a test for the latent heat of melting in the melt simple model at the moment, so it would be great if you could add one!

jdannberg commented on source/material_model/melt_simple.cc:
```diff
@@ -203,6 +205,7 @@ namespace aspect
     evaluate(const typename Interface<dim>::MaterialModelInputs &in, typename Interface<dim>::MaterialModelOutputs &out) const
     {
       ReactionRateOutputs<dim> *reaction_rate_out = out.template get_additional_output<ReactionRateOutputs<dim>>();
+      //this->melt_simple_fraction = std::make_unique<MeltModel::MeltSimpleFraction<dim>>();
```
You can remove this line.

jdannberg commented on source/material_model/melt_model/melt_simple_fraction.cc:
```diff
@@ -0,0 +1,205 @@
+  {
+    namespace MeltModel
+    {
+
```
Can you remove this empty line?

jdannberg reviewed (commented):
In general, this looks good to me! Thank you for modularizing the code!

Looking through the code, I think you need to make the entropy_change function part of your melting model as well, since it needs all of the same parameters, and you're now only reading them in in the melting model (and it would not really make sense to read them in twice). 

Perspectively, it would also make sense to think about what else you would need in the visco-plastic model. If you only need the melt fractions function, this is completely fine. If you want to track the melt fraction in a compositional field and/or add melt transport, you would also need all of the code in lines 230-314, which computes the actual reaction rates.

naliboff reviewed (commented):
@Grant-Block - Nice work, I think this in generally in very good shape! At this stage, I have just a few high level questions, but would like to take a more detailed look again after any renaming scheme updates and fixing the indentation.

General questions:
1. I think we should rename the new folder with the melting model from `melt_model` to `reaction_model`.  We might also want to consider changing the namespace from `MeltModel` to `ReactionModel` to be consistent with how the rheology and equation of state models are structured.
2. I would rename this PR to something like "Create melt reaction model modules"

@jdannberg - Thoughts on naming in point 1?

Grant-Block commented:
@naliboff, @jdannberg I've addressed all the above comments. Let me know if the name changes look good. My next step will be to get this module into the visco_plastic model, which @naliboff pointed out to me can be done through the reactive_fluid_transport model composited with visco plasticity. I'll start taking a look at how to do that now, can we meet to talk about that implementation today?

naliboff reviewed (approved):
@Grant-Block - Looks good to me, thanks!

jdannberg reviewed (commented):
Just two small changes, otherwise this looks good to me! Thank you!

jdannberg commented:
/rebuild
